### ISSUE TITLE
Theoretical new hex library for sharing command logic between hex and rebar3_hex (and others).

### ISSUE THREAD

**Issue by starbelly** [erlef/eef-build-and-packaging-wg#6]
Discuss the possibility of a new hex library to share command logic, such as hex.outdated, between hex, rebar3_hex, and hex related libraries and/or plugins. This requires enumerating the differences between the two as is and enumerating what's missing from rebar3_hex and other hex related libraries where it might be beneficial to share code between projects in particularly where it's the logic is a bit complicated, but also with the idea that getting new commands into plugins such as rebar3_hex in a more expedient fashion.

Potentials items that could belong to such a lib :

 - prompts / prompt logic : (i.e., prompting for usernames, passwords, etc.) This already pretty much the same, hex and rebar3_hex also uses the exact same implementation for getting passwords.

 -   formatters : this one is a bit eh, but I can tell you as part of a PR I'll be opening for rebar3_hex soon, it will have the same implementation for formatting generic errors from hex.

 - crypto - this one perhaps makes the most sense to share.

**Comment by ericmj** [comment]
Any ideas on how this would be implemented regarding the differences between the environments we run in, mix and rebar? Prompts and error output works differently in the two build tools so I am not sure how logic could be shared.

**Comment by starbelly** [comment]
> Any ideas on how this would be implemented regarding the differences between the environments we run in, mix and rebar? Prompts and error output works differently in the two build tools so I am not sure how logic could be shared.

That I'm not sure of, what would be the side effect of using `io:format` from mix? Also, if there is a problem perhaps we could pass callbacks around, but then maybe we're trying to hard to share a thing at this point.

**Comment by ericmj** [comment]
It would be a major refactor to change how hex prints, prompts, and outputs errors. We use things like `Mix.Shell` and `Mix.raise` because it integrates with mix and makes it easier to test. Looking solely from hex's point of view it would be a negative to move away from this. Another note is that hex and rebar3_hex prompts for passwords differently because of the different IO modes we run in.

**Comment by starbelly** [comment]
@ericmj Yeah, mix would have to case on errors and `Mix.raise`, etc. etc. A major refactor is not worth such a little amount of code. It's good enough to standardize on the UX. If we can agree on that I'm happy to do a PR to both rebar3_hex and mix on this.

**Comment by ericmj** [comment]
By standardize on UX, do you mean only agreeing on how we should structure or actually sharing code for it? I have a hard time seeing how code can be shared since the foundations are so different.

**Comment by michalmuskala** [comment]
I think part of an effort like this should be adding a function to OTP for properly reading passwords from the shell. The current solutions are rather hacky and sub-optimal. A native solution could be much more reliable.

**Comment by starbelly** [comment]
Not sharing code, simply making a more unified experience from the users perspective. This is something rebar3_hex has been moving towards. This is issue is around registration UX as an example is here : https://github.com/tsloughter/rebar3_hex/pull/189

 As for sharing code, I do think there might be some spots we can do that, but I wouldn't want to force it. Dependencies come at a cost even when people are working close together. I would definitely say at this point prompts and so forth are off the table. 

crypto related logic possible, but it's still so small it's hard for me to justify. 

Same with formatter(s). Though, I think some more formatter functions could be added to hex_core. Such as https://github.com/hexpm/hex/blob/1328d1edcaf0b85698f67f336dac317e641d5a3a/lib/hex/utils.ex#L179

The whole idea for me here is not about sharing code so much, but mostly providing the same output and UX. 

Original issue: 

> Discuss the possibility of a new hex library to share command logic, such as hex.outdated, between hex, rebar3_hex, and hex related libraries and/or plugins. This requires enumerating the differences between the two as is and enumerating what's missing from rebar3_hex and other hex related libraries where it might be beneficial to share code between projects in particularly where it's the logic is a bit complicated, but also with the idea that getting new commands into plugins such as rebar3_hex in a more expedient fashion.

I don't think this is a solid goal at this point. I don't think it's going to get anything out quicker and there's really not a lot of logic at this point that's so complicated that it warrants its own lib. Further, it's only valuable if both hex and rebar3_hex uses it, if not, then it misses the point. 

So, I think you're right, things are so different and small that this issue should probably be retired. It could also be re-opened at a later time if needed.

**Comment by starbelly** [comment]
> I think part of an effort like this should be adding a function to OTP for properly reading passwords from the shell. The current solutions are rather hacky and sub-optimal. A native solution could be much more reliable.

Yeah, I started looking into that... thought it was going to be simple, but think I was wrong. I need to revisit that in the next few days. Possibly just ditching the hacky solution that currently is available in OTP in the old shell and looking more into making native solutions work.  I was going to add the equivalent to the user drv, but that's where I started to have problems. Specifically I'd have to do some blocking and I believe that's antithetical to the idea behind `user`.

 This is the original problem we ran into in rebar3_hex and switched to the way hex was doing it for all platforms FWIW : https://bugs.erlang.org/browse/ERL-1069 ... Are you suggesting making this work or perhaps a nif?
